vm/src/stdlib/ssl.rs

```diff
@@ -0,0 +1,877 @@
+struct PySslContext {
+    ctx: RefCell<SslContextBuilder>,
+    check_hostname: bool,
+}
+
+impl fmt::Debug for PySslContext {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.pad("_SSLContext")
+    }
+}
+
+impl PyValue for PySslContext {
+    fn class(vm: &VirtualMachine) -> PyClassRef {
+        vm.class("_ssl", "_SSLContext")
+    }
+}
+
+#[pyimpl(flags(BASETYPE))]
+impl PySslContext {
+    fn builder(&self) -> RefMut<SslContextBuilder> {
+        self.ctx.borrow_mut()
+    }
+    fn ctx(&self) -> Ref<ssl::SslContextRef> {
+        Ref::map(self.ctx.borrow(), |ctx| unsafe {
+            &**(ctx as *const SslContextBuilder as *const ssl::SslContext)
+        })
+    }
+    fn ptr(&self) -> *mut sys::SSL_CTX {
+        self.ctx.borrow().as_ptr()
+    }
+
+    #[pyslot]
+    fn tp_new(cls: PyClassRef, proto_version: i32, vm: &VirtualMachine) -> PyResult<PyRef<Self>> {
+        let proto = SslVersion::try_from(proto_version)
+            .map_err(|_| vm.new_value_error("invalid protocol version".to_owned()))?;
+        let method = match proto {
+            SslVersion::Ssl2 => todo!(),
+            SslVersion::Ssl3 => todo!(),
+            SslVersion::Tls => ssl::SslMethod::tls(),
+            SslVersion::Tls1 => todo!(),
+            // TODO: Tls1_1, Tls1_2 ?
+            SslVersion::TlsClient => unsafe { ssl::SslMethod::from_ptr(sys::TLS_client_method()) },
+            SslVersion::TlsServer => unsafe { ssl::SslMethod::from_ptr(sys::TLS_server_method()) },
+        };
+        let mut builder =
+            SslContextBuilder::new(method).map_err(|e| convert_openssl_error(vm, e))?;
+
+        let check_hostname = proto == SslVersion::TlsClient;
+        builder.set_verify(if check_hostname {
+            SslVerifyMode::PEER | SslVerifyMode::FAIL_IF_NO_PEER_CERT
+        } else {
+            SslVerifyMode::NONE
+        });
+
+        let mut options = SslOptions::ALL & !SslOptions::DONT_INSERT_EMPTY_FRAGMENTS;
+        if proto != SslVersion::Ssl2 {
+            options |= SslOptions::NO_SSLV2;
+        }
+        if proto != SslVersion::Ssl3 {
+            options |= SslOptions::NO_SSLV3;
+        }
+        options |= SslOptions::NO_COMPRESSION;
+        options |= SslOptions::CIPHER_SERVER_PREFERENCE;
+        options |= SslOptions::SINGLE_DH_USE;
+        options |= SslOptions::SINGLE_ECDH_USE;
+        builder.set_options(options);
+
+        let mode = ssl::SslMode::ACCEPT_MOVING_WRITE_BUFFER | ssl::SslMode::AUTO_RETRY;
+        builder.set_mode(mode);
+
+        unsafe { sys::SSL_CTX_set_post_handshake_auth(builder.as_ptr(), 0) };
+
+        builder
+            .set_session_id_context(b"Python")
+            .map_err(|e| convert_openssl_error(vm, e))?;
+
+        PySslContext {
+            ctx: RefCell::new(builder),
+            check_hostname,
+        }
+        .into_ref_with_type(vm, cls)
+    }
+
+    #[pymethod]
+    fn set_ciphers(&self, cipherlist: CString, vm: &VirtualMachine) -> PyResult<()> {
```

Comment:
is there some advantage by using `CString` here instead of `String`?

Reply:
Oh, it'll return a value error if there's a nul character in the string, but the openssl set_ciphers implementation internally makes a new cstring and panics if it's not valid. I'll add a comment about it.

Reply:
I changed it to just manually check for a '\0' in a PyStringRef.